server/broadcast.go

```diff
@@ -434,44 +427,42 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string,
 			// preemptively fetch segments, they could be reading tampered
 			// data. Not an issue if the delivery protocol is being obeyed.
 			segHashLock.Lock()
-			segURLs[i] = url
+			tr.url[i] = url
 			segHashLock.Unlock()
 
 			if monitor.Enabled {
 				monitor.TranscodedSegmentAppeared(nonce, seg.SeqNo, sess.Profiles[i].Name)
 			}
 		}
 
+		wg.Add(len(res.Segments))
 		for i, v := range res.Segments {
+			tr.url[i] = v.Url
 			go dlFunc(v.Url, v.Pixels, i)
 		}
 
-		cond.L.Lock()
-		for n != 0 {
-			cond.Wait()
-		}
-		cond.L.Unlock()
+		wg.Wait()
 		if dlErr != nil {
-			return dlErr
+			return nil, dlErr
 		}
 
 		if verifier != nil {
-			err := verify(verifier, cxn, sess, seg, res.TranscodeData, segURLs)
+			err := verify(verifier, cxn, sess, seg, res.TranscodeData, tr.url)
 			if err != nil {
 				glog.Errorf("Error verifying nonce=%d manifestID=%s seqNo=%d err=%s", nonce, cxn.mid, seg.SeqNo, err)
-				return err
+				return tr, err
 			}
 		}
 
-		for i, url := range segURLs {
+		for i, url := range tr.url {
 			err = cpl.InsertHLSSegment(&sess.Profiles[i], seg.SeqNo, url, seg.Duration)
 			if err != nil {
 				// We assume playlist insertion is *not* recoverable for now
 				glog.Errorf("Playlist insertion error nonce=%d manifestID=%s seqNo=%d err=%s", nonce, cxn.mid, seg.SeqNo, err)
 				if monitor.Enabled {
 					monitor.SegmentTranscodeFailed(monitor.SegmentTranscodeErrorPlaylist, nonce, seg.SeqNo, err, false)
 				}
-				return nil
+				return tr, nil
```

Comment:
Is there a reason we return tr here but not for other errors?

Reply:
Actually, after thinking about this more, I think we should just ignore error returned from `InsertHLSSegment`. Only error it can return is `segment already exists`. Previously `InsertHLSSegment` was called right after downloading segment, independently for each segment. So, in case when one rendition was successfully downloaded and inserted into playlist, but download of second rendition resulted in error and caused to re-try transcoding, first rendition was inserted into playlist again and was returning error. But we was ignoring it anyway.
I'll remove check for this error.

Reply:
Probably OK for now. I wonder if we'll ever have other types of errors, but agreed that how it currently works makes things a bit hard to handle. Maybe add a comment explaining the behavior during retries.

This would be another unrelated change, so best to put that in a separate commit prior to the change to `transcodeSegment`.

Reply:
> Maybe add a comment explaining the behavior during retries.

But after your change there is no special behaviour - insert into playlist is atomic regarding each segment - either every segments are transcoded and inserted, or none - so we really shouldn't hit this error.
But I can can check for this error and report it to log and grafana, but just ignore - this way, if we will hit it, we will know that something is deeply wrong, probably in other part of the stack.

Reply:
fixed